### ISSUE TITLE
Clients read config from file resulting in EMFILE (too many open files) errors

### ISSUE THREAD

**Issue by moltar** [aws/aws-sdk-js-v3#3019]
## Describe the bug

This applies to **every** auto-generated client, but I will use `client-sqs` as an example.

This is related to: #2271, #2027, #2993

---

Note that in the above snippet, `loadNodeConfig` is called multiple times to get the defaults for the client configuration:

https://github.com/aws/aws-sdk-js-v3/blob/bda3b40dd773511b8d5c84a07e9e158a70073523/clients/client-sqs/src/runtimeConfig.ts#L31-L53

Which is the exported fn `loadConfig` from `@aws-sdk/node-config-provider` package:

https://github.com/aws/aws-sdk-js-v3/blob/73923416ef2beaefbe7feda0473bc39c734be4b8/packages/node-config-provider/src/configLoader.ts#L27-L37

Which then uses `fromSharedConfigFiles` to load config values from the disk.

This is a very undesirable "feature", especially in the serverless environments.

Because under heavy load this results in the following errors:

```
    error: {
      "type": "NodeError",
      "message": "A system error occurred: uv_os_homedir returned EMFILE (too many open files)",
      "stack":
          SystemError [ERR_SYSTEM_ERROR]: A system error occurred: uv_os_homedir returned EMFILE (too many open files)
              at Object.getHomeDir (/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:82:17)
              at Object.loadSharedConfigFiles (/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:11:89)
              at null.<anonymous> (/node_modules/@aws-sdk/node-config-provider/dist-cjs/fromSharedConfigFiles.js:9:53)
              at null.<anonymous> (/node_modules/@aws-sdk/property-provider/dist-cjs/chain.js:11:28)
              at runMicrotasks (<anonymous>)
              at processTicksAndRejections (internal/process/task_queues.js:95:5)
              at null.coalesceProvider (/node_modules/@aws-sdk/property-provider/dist-cjs/memoize.js:13:24)
              at Object.isConstant (/node_modules/@aws-sdk/property-provider/dist-cjs/memoize.js:24:28)
              at Object.getEndpointFromRegion (/node_modules/@aws-sdk/config-resolver/dist-cjs/endpointsConfig/utils/getEndpointFromRegion.js:12:34)
              at null.buildHttpRpcRequest (/node_modules/@aws-sdk/client-sqs/dist-cjs/protocols/Aws_query.js:2540:68)
      "code": "ERR_SYSTEM_ERROR",
      "info": {
        "errno": -24,
        "code": "EMFILE",
        "message": "too many open files",
        "syscall": "uv_os_homedir"
      },
      "errno": -24,
      "syscall": "uv_os_homedir"
    }
```

---

Yes the call is memoized, but if your Lambda is getting executed heavily, and installation happens within the handler, then this call happens multiple times.

#### SDK version number

`@aws-sdk/client-sqs@3.40.0`

#### Is the issue in the browser/Node.js/ReactNative?

Node.js

#### Details of the browser/Node.js/ReactNative version

`node -v` v14.17.5

## Steps to reproduce

```ts
import { SQS } from '@aws-sdk/client-sqs'

const sqs = new SQS({})
```

#### Observed behavior

```
    error: {
      "type": "NodeError",
      "message": "A system error occurred: uv_os_homedir returned EMFILE (too many open files)",
      "stack":
          SystemError [ERR_SYSTEM_ERROR]: A system error occurred: uv_os_homedir returned EMFILE (too many open files)
              at Object.getHomeDir (/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:82:17)
              at Object.loadSharedConfigFiles (/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:11:89)
              at null.<anonymous> (/node_modules/@aws-sdk/node-config-provider/dist-cjs/fromSharedConfigFiles.js:9:53)
              at null.<anonymous> (/node_modules/@aws-sdk/property-provider/dist-cjs/chain.js:11:28)
              at runMicrotasks (<anonymous>)
              at processTicksAndRejections (internal/process/task_queues.js:95:5)
              at null.coalesceProvider (/node_modules/@aws-sdk/property-provider/dist-cjs/memoize.js:13:24)
              at Object.isConstant (/node_modules/@aws-sdk/property-provider/dist-cjs/memoize.js:24:28)
              at Object.getEndpointFromRegion (/node_modules/@aws-sdk/config-resolver/dist-cjs/endpointsConfig/utils/getEndpointFromRegion.js:12:34)
              at null.buildHttpRpcRequest (/node_modules/@aws-sdk/client-sqs/dist-cjs/protocols/Aws_query.js:2540:68)
      "code": "ERR_SYSTEM_ERROR",
      "info": {
        "errno": -24,
        "code": "EMFILE",
        "message": "too many open files",
        "syscall": "uv_os_homedir"
      },
      "errno": -24,
      "syscall": "uv_os_homedir"
    }
```

#### Expected behavior

1. Do not error out
2. Do not read config from disk by default, or allow overriding this behaviour.

#### Screenshots

N/A

## Additional context

N/A

**Comment by moltar** [comment]
Added a `console.log` to `slurpFile`:

https://github.com/aws/aws-sdk-js-v3/blob/bda3b40dd773511b8d5c84a07e9e158a70073523/packages/shared-ini-file-loader/src/index.ts#L100

```ts
const slurpFile = (path) => new Promise((resolve, reject) => {
    console.log(path, new Error().stack)
```

And tested with the following two-liner file:

```ts
import { SQS } from '@aws-sdk/client-sqs'
const sqs = new SQS({})
```

## Results

```
/Users/user/.aws/config Error: 
    at /Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:66:23
    at new Promise (<anonymous>)
    at slurpFile (/Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:64:29)
    at Object.loadSharedConfigFiles (/Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:13:9)
    at defaultProvider (/Users/user/project/node_modules/@aws-sdk/credential-provider-node/dist-cjs/index.js:17:57)
    at Object.credentialDefaultProvider (/Users/user/project/node_modules/@aws-sdk/client-sts/dist-cjs/defaultRoleAssumers.js:10:68)
    at Object.resolveAwsAuthConfig (/Users/user/project/node_modules/@aws-sdk/middleware-signing/dist-cjs/configurations.js:10:17)
    at new SQSClient (/Users/user/project/node_modules/@aws-sdk/client-sqs/dist-cjs/SQSClient.js:20:48)
    at new SQS (/Users/user/project/node_modules/@aws-sdk/client-sqs/dist-cjs/SQS.js:25:1)
    at Object.<anonymous> (/Users/user/project/test/sdk.ts:3:13)
/Users/user/.aws/credentials Error: 
    at /Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:66:23
    at new Promise (<anonymous>)
    at slurpFile (/Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:64:29)
    at Object.loadSharedConfigFiles (/Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:14:9)
    at defaultProvider (/Users/user/project/node_modules/@aws-sdk/credential-provider-node/dist-cjs/index.js:17:57)
    at Object.credentialDefaultProvider (/Users/user/project/node_modules/@aws-sdk/client-sts/dist-cjs/defaultRoleAssumers.js:10:68)
    at Object.resolveAwsAuthConfig (/Users/user/project/node_modules/@aws-sdk/middleware-signing/dist-cjs/configurations.js:10:17)
    at new SQSClient (/Users/user/project/node_modules/@aws-sdk/client-sqs/dist-cjs/SQSClient.js:20:48)
    at new SQS (/Users/user/project/node_modules/@aws-sdk/client-sqs/dist-cjs/SQS.js:25:1)
    at Object.<anonymous> (/Users/user/project/test/sdk.ts:3:13)
/Users/user/.aws/config Error: 
    at /Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:66:23
    at new Promise (<anonymous>)
    at slurpFile (/Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:64:29)
    at Object.loadSharedConfigFiles (/Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:13:9)
    at /Users/user/project/node_modules/@aws-sdk/node-config-provider/dist-cjs/fromSharedConfigFiles.js:9:53
    at /Users/user/project/node_modules/@aws-sdk/property-provider/dist-cjs/chain.js:11:28
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at coalesceProvider (/Users/user/project/node_modules/@aws-sdk/property-provider/dist-cjs/memoize.js:13:24)
    at isConstant (/Users/user/project/node_modules/@aws-sdk/property-provider/dist-cjs/memoize.js:24:28)
/Users/user/.aws/credentials Error: 
    at /Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:66:23
    at new Promise (<anonymous>)
    at slurpFile (/Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:64:29)
    at Object.loadSharedConfigFiles (/Users/user/project/node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js:14:9)
    at /Users/user/project/node_modules/@aws-sdk/node-config-provider/dist-cjs/fromSharedConfigFiles.js:9:53
    at /Users/user/project/node_modules/@aws-sdk/property-provider/dist-cjs/chain.js:11:28
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at coalesceProvider (/Users/user/project/node_modules/@aws-sdk/property-provider/dist-cjs/memoize.js:13:24)
    at isConstant (/Users/user/project/node_modules/@aws-sdk/property-provider/dist-cjs/memoize.js:24:28)
```

As you can see, it is trying to read the same files twice, and makes at least 4 IO operations per instance of a client.

**Comment by moltar** [comment]
Now, making this a bit more realistic and instantiating SQS client in a loop.

This would be the case when it is instantiated inside a lambda handler.

Well, it wouldn't be a loop, but it would be instantiated multiple times when lambda is being executed warm, and very frequently.

```ts
import { SQS } from '@aws-sdk/client-sqs'

for (const iter of [1, 2]) {
  const sqs = new SQS({})
}
```

Then:

```
ts-node test/sdk.ts | grep Error | wc -l
```

## Results

In 8 executions!

So the results are never cached globally, and are simply cached per client instance.

**Comment by moltar** [comment]
Here's a way to remove unnecessary calls:

```ts
import { SQS } from '@aws-sdk/client-sqs'

const sqs = new SQS({
  // removes first 2 calls
  credentials: {
    accessKeyId: 'x',
    secretAccessKey: '',
  },

  // removes second 2 calls
  defaultUserAgentProvider: async () => [],
})

```

**Comment by moltar** [comment]
But wait ... there is more!

If you actually try to use the methods of the queue, this will result in even more calls!

```ts
import { SQS } from '@aws-sdk/client-sqs'

const sqs = new SQS({
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID || '',
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY || '',
    sessionToken: process.env.AWS_SESSION_TOKEN || '',
  },
  region: 'us-east-1',
  defaultUserAgentProvider: async () => [],
})

async function main() {
  await sqs.listQueues({})
}

main()
```

## Results

```
$> ts-node test/sdk.ts | grep Error | wc -l

8
```

So we get 8 IO calls for each method call (I tried `sendMessage` too, it's the same)!

This is ridiculous and needs to be addressed immediately, as it is causing major issues!

**Comment by moltar** [comment]
Interestingly, calling `listQueues` in a loop, does not cause additional calls. So, I guess the results _do_ get cached.

**Comment by moltar** [comment]
If anyone stumbles on this, the only way I found to shut the file reading, is to do the following hack.

**Note**: It must proceed other `@aws-sdk/*` imports.

```ts
import * as sharedIniFileLoader from '@aws-sdk/shared-ini-file-loader'

Object.assign(sharedIniFileLoader, {
  loadSharedConfigFiles: async (): Promise<sharedIniFileLoader.SharedConfigFiles> => ({
    configFile: {},
    credentialsFile: {},
  }),
})
```

**Comment by petermorlion** [comment]
I can confirm that @moltar 's workaround solves this issue in AWS Lambda functions.

**Comment by adrai** [comment]
The aws-sdk should detect if running in lambda and omit the loadSharedConfigFiles call completely, or at least provide a "supported" way to disable the loadSharedConfigFiles call...

**Comment by adrai** [comment]
Without this workaround suggested by @moltar it seems v3.49.0 is even getting worse, and opening more file descriptors: probably because of this https://github.com/aws/aws-sdk-js-v3/pull/3192 ?
//cc @AllanZhengYP @trivikr

**Comment by adrai** [comment]
```js
import { DynamoDBClient } from '@aws-sdk/client-dynamodb'
import fs from 'fs'
import { lsof } from 'list-open-files'

let readFileCount = 0
const originalReadFile = fs.readFile
fs.readFile = function (f) {
  // console.log(`reading file: ${f}`)
  readFileCount++
  return originalReadFile.apply(fs, arguments)
}

const clientCreations = 1000
for (let index = 0; index < clientCreations; index++) {
  new DynamoDBClient()
}

console.log(`made ${clientCreations} client instances`)
console.log(`called readFile ${readFileCount} times`)

lsof().then((ret) => {
  const openFiles = ret[0].files.length
  console.log(`${openFiles} open fds`)

  lsof().then((ret) => {
    const openFiles = ret[0].files.length
    console.log(`${openFiles} open fds after a bit later`)
  })
})
```

```
made 1000 client instances
called readFile 2000 times
4030 open fds
30 open fds after a bit later
```

**Comment by ffxsam** [comment]
@AllanZhengYP @ajredniwja Could you please check into this when you get a chance? This is a serious issue that impacted our production system recently.

**Comment by adrai** [comment]
@vudh1 you assigned this issue to yourself, can you say something about it?

**Comment by adrai** [comment]
fyi: found an interesting package: https://github.com/samswen/lambda-emfiles

made some tests:

Some insights (same workload for all tests)

v3.46.0 without any workarounds:
![image](https://user-images.githubusercontent.com/1086194/152322220-c12efba7-6eea-41d4-b8a7-f25fa4f56958.png)
![image](https://user-images.githubusercontent.com/1086194/152324433-1fa5bb11-e18d-455c-a0aa-24919bc13e9f.png)
=> some leaks (around 230 emfiles)

v3.49.0 without any workarounds:
![image](https://user-images.githubusercontent.com/1086194/152322285-081469dd-6327-4559-819e-bcd9741a8bf2.png)
![image](https://user-images.githubusercontent.com/1086194/152324356-dab148b1-1660-4e7e-9301-0f89a14b50e1.png)
=> some leaks and more emfiles (up to more than 600 emfiles)

v3.49.0 with loadSharedConfigFiles workaround:
![image](https://user-images.githubusercontent.com/1086194/152323613-f023b27f-7f6b-4499-9185-c2e46bedd2fd.png)
![image](https://user-images.githubusercontent.com/1086194/152323032-1ab24477-d13e-42a3-8086-27316ab37ff5.png)
=> much less leaks, probably also less emfiles (up to more than 400 emfiles)

v3.46.0 with loadSharedConfigFiles workaround:
![image](https://user-images.githubusercontent.com/1086194/152324198-989aae0f-3165-4b03-b1b2-9daf0ba1996f.png)
![image](https://user-images.githubusercontent.com/1086194/152323953-d7d54229-308e-4e67-8c62-1366e6aa6292.png)
=> much less leaks, much less emfiles (up to 180 emfiles)


**So seems v3.46.0 with loadSharedConfigFiles workaround is the best**

**Comment by adrai** [comment]
An additional curiosity:

Defining a custom requestHandler, with a very low socketTimeout reduces drastically the emfiles count:

```js
  requestHandler: new NodeHttpHandler({
      httpsAgent,
      socketTimeout: 1000, // <- this decreases the emfiles count
      connectionTimeout: 6000
    })
```

**Comment by moltar** [comment]
> An additional curiosity:
> 
> Defining a custom requestHandler, with a very low socketTimeout reduces drastically the emfiles count:
> 
> ```js
>   requestHandler: new NodeHttpHandler({
>       socketTimeout: 10000 // <- this decreases the emfiles count, the Node.js default is 120000
>   })
> ```

Great find! Any idea what the default value is?

**Comment by moltar** [comment]
Wondering if this timeout perhaps need to equal Lambda timeout? Maybe this can be auto-set based on the `getRemainingTimeInMillis`?

https://docs.aws.amazon.com/lambda/latest/dg/nodejs-context.html

**Comment by moltar** [comment]
The only downside with this approach would be that the client has to be instantiated inside the handler. And I usually like to have them outside of the handler to preserve the instances across invocations.

**Comment by adrai** [comment]
As this seems to be the timeout for an idle socket, I think in lambda this could be automatically defaulted to a very low value, at least magnitute of 10... so 12 seconds instead of 2 minutes... but even lower so 1 second for example, should work.

Would be nice if some AWS internal engineer would help to argument on this topic.

**Comment by ffxsam** [comment]
This issue has been open for nearly three months and no one from AWS has chimed in. I'll reach out to my account manager and see if they can help move things along here.

**Comment by adrai** [comment]
> This issue has been open for nearly three months and no one from AWS has chimed in. I'll reach out to my account manager and see if they can help move things along here.

https://twitter.com/trivikram/status/1489588283726204928?s=21

**Comment by ffxsam** [comment]
Thanks for that update!

**Comment by trivikr** [comment]
AWS SDK for JavaScript team discussed this issue in our scrum today.
We're evaluating adding some kind of lock/mutex to `fromSharedConfigFiles` function, and I'll provide an update here.

**Comment by trivikr** [comment]
Thank you @moltar for the detailed repro for the issue, and providing code samples.
I verify that there are four async calls to readFile in one instantiation of the client.

<details>
<summary>Code</summary>

```js
import { S3 } from "@aws-sdk/client-s3"; // v3.49.0

console.log("Before client instantiation");
const client = new S3({});
console.log("After client instantiation");
```

</details>

<details>
<summary>node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js</summary>

```console
Before client instantiation
start: readFile call #1 from slurpFile
start: readFile call #2 from slurpFile
After client instantiation
start: readFile call #3 from slurpFile
start: readFile call #4 from slurpFile
end: readFile call #1 from slurpFile
end: readFile call #2 from slurpFile
end: readFile call #3 from slurpFile
end: readFile call #4 from slurpFile
```

</details>

**Comment by trivikr** [comment]
I verified that readFile is called twice for `~/.aws/config` and twice for `~/.aws/credentials`

<details>
<summary>node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js</summary>

```js
...
let readFileCallTotal = 0;
const slurpFile = (path) => new Promise((resolve, reject) => {
    console.log(`readFile call #${++readFileCallTotal} from slurpFile\n    for path: ${path}`);
    fs_1.readFile(path, "utf8", (err, data) => {
...
```

</details>

<details>
<summary>Output</summary>

```console
Before client instantiation
readFile call #1 from slurpFile
    for path: /home/trivikr/.aws/config
readFile call #2 from slurpFile
    for path: /home/trivikr/.aws/credentials
After client instantiation
readFile call #3 from slurpFile
    for path: /home/trivikr/.aws/config
readFile call #4 from slurpFile
    for path: /home/trivikr/.aws/credentials
```

</details>

**Comment by adrai** [comment]
> AWS SDK for JavaScript team discussed this issue in our scrum today.
> 
> We're evaluating adding some kind of lock/mutex to `fromSharedConfigFiles` function, and I'll provide an update here.

@trivikr The leaking file descriptors of the readFile calls is just one part... the file descriptors of the http requests are the other part of the issue, which could be mitigated with a lower socketTimeout value.

**Comment by trivikr** [comment]
I can verify that readFile is called 8 times when two clients are instantiated.

<details>
<summary>Code</summary>

```js
import { S3 } from "@aws-sdk/client-s3"; // v3.49.0

console.log("Before client1 instantiation");
const client1 = new S3({});
console.log("After client1 instantiation");

console.log("Before client2 instantiation");
const client2 = new S3({});
console.log("After client2 instantiation");
```

</details>

<details>
<summary>node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js</summary>

```js
...
let readFileCallTotal = 0;
const slurpFile = (path) => new Promise((resolve, reject) => {
    console.log(`readFile call #${++readFileCallTotal} from slurpFile`);
    fs_1.readFile(path, "utf8", (err, data) => {
...
```

</details>

<details>
<summary>Output</summary>

```console
Before client1 instantiation
readFile call #1 from slurpFile
readFile call #2 from slurpFile
After client1 instantiation
Before client2 instantiation
readFile call #3 from slurpFile
readFile call #4 from slurpFile
After client2 instantiation
readFile call #5 from slurpFile
readFile call #6 from slurpFile
readFile call #7 from slurpFile
readFile call #8 from slurpFile
```

</details>

**Comment by trivikr** [comment]
> The leaking file descriptors of the readFile calls is just one part... the file descriptors of the http requests are the other part of the issue, which could be mitigated with a lower socketTimeout value.

@adrai Can you create a new [bug report](https://github.com/aws/aws-sdk-js-v3/issues/new?assignees=&labels=bug%2C+needs-triage&template=---bug-report.md&title=)? It would be easy for tracking fixes.

**Comment by trivikr** [comment]
I can verify that readFile is called 16 times with one client instantiation and one operation call

<details>
<summary>Code</summary>

```js
import { S3 } from "@aws-sdk/client-s3"; // v3.49.0

console.log("Before client instantiation");
const client = new S3({});
console.log("After client instantiation");

console.log("Before calling operation");
await client.listBuckets({});
console.log("After calling operation");
```

</details>

<details>
<summary>node_modules/@aws-sdk/shared-ini-file-loader/dist-cjs/index.js</summary>

```js
...
let readFileCallTotal = 0;
const slurpFile = (path) => new Promise((resolve, reject) => {
    console.log(`readFile call #${++readFileCallTotal} from slurpFile`);
    fs_1.readFile(path, "utf8", (err, data) => {
...
```

</details>

<details>
<summary>Output</summary>

```console
Before client instantiation
readFile call #1 from slurpFile
readFile call #2 from slurpFile
After client instantiation
Before calling operation
readFile call #3 from slurpFile
readFile call #4 from slurpFile
readFile call #5 from slurpFile
readFile call #6 from slurpFile
readFile call #7 from slurpFile
readFile call #8 from slurpFile
readFile call #9 from slurpFile
readFile call #10 from slurpFile
readFile call #11 from slurpFile
readFile call #12 from slurpFile
readFile call #13 from slurpFile
readFile call #14 from slurpFile
readFile call #15 from slurpFile
readFile call #16 from slurpFile
After calling operation
```

</details>

**Comment by adrai** [comment]
> > The leaking file descriptors of the readFile calls is just one part... the file descriptors of the http requests are the other part of the issue, which could be mitigated with a lower socketTimeout value.
> 
> @adrai Can you create a new [bug report](https://github.com/aws/aws-sdk-js-v3/issues/new?assignees=&labels=bug%2C+needs-triage&template=---bug-report.md&title=)? It would be easy for tracking fixes.

here we go ;-) https://github.com/aws/aws-sdk-js-v3/issues/3279

**Comment by ffxsam** [comment]
Pardon any confusion on my part, but it seems like the file descriptor issue is happening for every single API call, not just upon instantiation. I'm only instantiating `S3Client` once, but then I call `CopyObject` 2000 times, and I get the `EMFILE` error. Does this match other people's experiences?

**Comment by adrai** [comment]
> Pardon any confusion on my part, but it seems like the file descriptor issue is happening for every single API call, not just upon instantiation. I'm only instantiating `S3Client` once, but then I call `CopyObject` 2000 times, and I get the `EMFILE` error. Does this match other people's experiences?

That's why I created #3279 😉

**Comment by trivikr** [comment]
Hi folks, I've posted WIP PR to remove concurrent/duplicate calls in slurpFile at https://github.com/aws/aws-sdk-js-v3/pull/3281

Do post your comments if you take a look.
cc folks who had commented on this issue: @adrai @ffxsam @moltar @petermorlion

**Comment by ffxsam** [comment]
@trivikr Thanks, Trivikram, much appreciated!

**Comment by moltar** [comment]
> Hi folks, I've posted WIP PR to remove concurrent/duplicate calls in slurpFile at ~~https://github.com/aws/aws-sdk-js-v3/pull/3281~~ https://github.com/aws/aws-sdk-js-v3/pull/3282
> 
> 
> 
> Do post your comments if you take a look.
> 
> cc folks who had commented on this issue: @adrai @ffxsam @moltar @petermorlion 

Thank you for looking at the issue.

But why do we need to have any calls in the lambda environment?

We need some way to disable config reading completely.

Thank you.

**Comment by trivikr** [comment]
> Hi folks, I've posted WIP PR to remove concurrent/duplicate calls in slurpFile at ~#3281~ #3282

Since I spent too much time on this issue, I refactored shared-ini-file-loader into multiple components and posted an update in https://github.com/aws/aws-sdk-js-v3/pull/3285

The fix is a merge of both #3281 and #3282:
* I'd to use hash and promise callback queue on slurpFile so that hash can be specific to path, and not a mix of configPath+credsPath
* The lastModified key from hash is removed, as data is anyway memoized downstream and readFile will be called for any path just once.

**Comment by trivikr** [comment]
> But why do we need to have any calls in the lambda environment?
> We need some way to disable config reading completely.

@moltar Can you create a [feature request](https://github.com/aws/aws-sdk-js-v3/issues/new?assignees=&labels=feature-request&template=---feature-request.md&title=) for this proposal?

**Comment by moltar** [comment]
I feel this is part of the same issue, no?

The client should not check the disk for credentials if there are env variables set, which is always the case in the Lambda env.

I can see it still trying to load `~/.aws/config` with env vars present though, which _might_ be valid, in a weird way. E.g. if someone puts the config there during Lambda execution, before SDK instantiation, for whatever reason.

I think one approach would be to continue make use the provider pattern, as it is already applied. Just make sure the SDK clients respect any explicit providers given. E.g. could be `ConfigProvider` and `CredentialsProvider`, when supplying both of those with instances of (_imaginary_) `ConfigFromMemoryProvider` and `CredentialsFromEnvironmentProvider` then it'd avoid making any disk IO and use the provider _only_.

If you think this is a viable approach, I will open a separate issue.

**Comment by manchicken** [comment]
I'm still seeing this issue in 3.52.0.

**Comment by adrai** [comment]
> I'm still seeing this issue in 3.52.0.

It's probably caused by the sdk requests: #3279

**Comment by trivikr** [comment]
> @adrai totally possible. I'm trying @moltar's work-around right now to see if that helps.

The workaround in https://github.com/aws/aws-sdk-js-v3/issues/3019#issuecomment-966900587 will remove the remaining two readFile calls per client, but the https://github.com/aws/aws-sdk-js-v3/issues/3279 will still remain.

Please use low socket timeouts as suggested by @adrai

**Comment by kellertk** [comment]
Reopening this issue per https://twitter.com/thdxr/status/1696615147295379586

**Comment by trivikr** [comment]
From the tweet

> in terms of areas of improvements with these things it's typically death by a thousand cut - the current struggle i'm having is this issue that's been closed and unclear where it landed

It's very clear that the issue discussed in this specific report was resolved in https://github.com/aws/aws-sdk-js-v3/pull/3285
The testing section of the PR shows how number of readFile calls are reduced from 4/8/16 to 2 calls in different scenarios.

I also verified that another issue created as a follow-up was also resolved in https://github.com/aws/aws-sdk-js-v3/issues/3279#issuecomment-1105517410, where we resolve config provider only once.

**Comment by trivikr** [comment]
If you're coming across EMFILE or any other issues in your setup, please create a new [feature request](https://github.com/aws/aws-sdk-js-v3/issues/new?assignees=&labels=feature-request&template=---feature-request.md&title=) or upvote [existing ones](https://github.com/search?q=repo%3Aaws%2Faws-sdk-js-v3+EMFILE&type=issues) if they're already reported.
